notes/overriding.md

```diff
@@ -0,0 +1,797 @@
+function-based scheme as the NixOS module system. One big difference with the
+NixOS module system is that it would be built-in in the language, being usable
+for any native record whatsoever, and having good interaction with the rest of
+the language features. The implementation would have first class access to
+locations, to the AST, the memory layout, the metadata, and so on.
+
+We first describe the ideas at a high-level. Then, we review the issue and
+challenges of this approach, and propose concrete solutions to overcome or
+mitigate them. Finally, technical operational semantics are laid out in a
+dedicated section.
+
+### Recursive records & merging
+
+**This section defines the semantics, rather than an actual efficient
+implementation**.
+
+As before, it is useful to see recursive records represented as functions - or
+constructors - that take a `self` parameter and return a final non recursive record,
+in the same way as the original overriding mechanism of Nixpkgs (the non-existing
+syntax `def := value` is used to insist on the fact that we are defining new objects):
+
+```
+r = {
+  a = 1;
+  b = a + 1;
+}
+// Definition of the representation of r
+repr(r) := fun self => {
+  a = 1;
+  b = self.a + 1;
+}
+```
+
+Field access amounts to compute a fix-point:
+
+```
+repr(r).foo := let fix = repr(r) fix in fix.foo
+```
+
+Merging simply merges the underlying representations:
+
+```
+repr(r1) & repr(r2) := fun self => r1 self & r2 self
+```
+
+### Priorities
+
+Merge is fundamentally commutative. The problem is, not all fields should be
+treated the same: some are default values that ought to be overridden, some are
+high-priority values that ought to override. The solution is to use a priority
+system, encoding the precedence of each value while retaining commutativity
+(only the priority annotation counts, not the order in which values are merged).
+
+Priorities have drawbacks. One is non-locality: the final result depends on
+other priorities of the field's values you are being merged with, potentially
+defined elsewhere. However, this can be mitigated by good defaults and an
+adapted set of priorities, such as a bottom `default`, a top `force`, and a
+infinite range of integers priorities in between. Doing so, it's easy to just
+use `default` or `force` to override or be overridden by anything, without
+knowing the precise integer priority. Integer priorities still gives freedom
+with an infinite supply of levels if required.
+
+### Custom merge functions
+
+Custom merge functions would be specified by a `merge` metavalue attribute. In
+order to enforce commutativity, they would receive their arguments in an
+indistinguishable order, such as having the type `{lower: Dyn, higher: Dyn,
+priority: <Different, Equal> -> Dyn`.  If both have the same priority, the order
+is not specified, and may even be randomized by the interpreter. The `priority`
+field indicates when it is the case, if this case needs special handling.
```

Comment:
This doesn't actually enforce commutativity, it just encourages it. The function can still be non-commutative, and this will introduce weird ordering dependencies that people might be surprised by. 

If the order is randomized, then something might work and break randomly half the time. That seems like a difficult bug to track down.

If the order is not randomized, then you may inadvertently create a dependency on order that only shows up when you refactor.

Reply:
To be clear, the way things are combined in NixOS modules is also non-commutative, it's just usually okay enough to get by. For instance, bash scripts are concatenated, which is only commutative if you're careful about what content you're putting in those bash scripts.

Reply:
The general problem is undecidable (not just by Rice's theorem, but in a deeper way, see e.g. [this resource](https://dl.acm.org/doi/10.1145/570886.570889)) and I don't think we can do much better (without a proof system to express and prove that a function is commutative). The problem is here for type class laws in Haskell for example, where you can use the do-notation with instances that don't actually respect monad laws, and could make expected equivalences invalid in practice.

One solution in the case of Nickel could be to use contracts to check that each application of the custom merge function respect commutativity:
```
let Commutative = fun label f x y =>
  let value = f x y in
  if value == f y x then
    value
  else
    contracts.blame label
```

That would add quite some overhead though. Although it will be possible to disable contracts with a flag in the future.

Reply:
Ooh, I haven't looked into contracts but that looks really cool. This seems to just be wrapping the function in a way that verifies it's commutative, which makes a lot of sense to me.

If that were how things were implemented, I would be much more sold on this solution. Even if it weren't enforced, but were expected that things were actually commutative. Like how monad laws are expected to hold. The person implementing the monad laws are on the hook for making sure they hold; it's not the job of the person *using* `bind` to make sure they pass in the right arguments so that the monad laws hold.

By contrast, NixOS modules have concatenation of bash scripts as a "commutative" operation. But that clearly isn't. I have to limit myself to a subset of bash that is commutative. But it's even worse than that, For instance at one point in my configuration I set the default cursor shape in the X init script. If I were to do that same operation (with different cursor shapes) in two different places that were supposed to be merged, the behaviour is undefined. It's more like if there were tons of examples of Monads people use on a daily basis where the monad laws only hold on a subset of that type and people are just expected to only use that subset in a `do` block.

Maybe this is a way that nickel would be different from NixOS modules (I don't know how cue handles this).

Reply:
> By contrast, NixOS modules have concatenation of bash scripts as a "commutative" operation. But that clearly isn't. I have to limit myself to a subset of bash that is commutative. But it's even worse than that, For instance at one point in my configuration I set the default cursor shape in the X init script. If I were to do that same operation (with different cursor shapes) in two different places that were supposed to be merged, the behaviour is undefined. It's more like if there were tons of examples of Monads people use on a daily basis where the monad laws only hold on a subset of that type and people are just expected to only use that subset in a `do` block.
> 
> Maybe this is a way that nickel would be different from NixOS modules (I don't know how cue handles this).

CUE doesn't handle this, because it doesn't have custom merge functions. You can only override a default value with a new value once, but not combine them in a specific way.

This is indeed a problem, but it looks like it is one with the current NixOS module implementation. I see two cases:

- Your merge function is morally commutative, but may not be in practice (for example, if you represent sets with lists, or a set of bash programs as a single string). One solution is to just not care, if we don't enforce commutativity with a contract, as long as it doesn't matter in the end (e.g. the tool consuming the config doesn't care bout it either). Or, you can make it actually commutative (use an ordered list without duplicated elements, for example).
- Your merge function is fundamentally not commutative. Then the merge function can use the priority information to decide of the order: it could fail to merge two scripts with the same priority for example, and otherwise concatenate them from the higher priority to the lower one. There are still subtleties though: you need to remember the priority to keep the operation commutative (otherwise merging priorities 4,2,1 and 4,1,2 wouldn't give the same thing).

To me the bash scripts problem seems to fall in the latter case: while the concatenation is morally commutative for a subset of inputs, it's simply not in general.

Reply:
Agreed. If there is a strong norm (or enforcement) that we don't have merge functions of the latter sort, we should be alright. clearly it can also work to have such functions, but I think it adds cognitive overhead.

As for the 4,2,1 =/= 4,1,2 issue, that's a failure of associativity I think? Is that an additional property we want to enforce / have as a norm?